contracts/Auction.sol

```diff
@@ -0,0 +1,183 @@
+        uint256 endBlock,
+        uint256 startPrice,
+        uint256 endPrice
+    ) internal {
+        console.log("starting action with endBlock: %s startPrice: %s endPrice: %s", endBlock, startPrice, endPrice);
+        require(isAuctionClosed() == true, "err:auction currently running");
+        _validateAuctionParams(endBlock, startPrice, endPrice);
+        _setAuctionPrivateMembers(endBlock, startPrice, endPrice);
+        _emitAuctionStart();
+        _isAuctionClosed = false;
+    }
+
+    function stopAuction() external onlyAuctioneer {
+        _isAuctionClosed = true;
+        _emitAuctionForcedStopped();
+    }
+
+    function getCurrentBiddingPrice() external view override returns (uint256 auctionBiddingPrice) {
+        /// If reached endBlock , handle auction that is "closed"
+        /// ELSE calculate current price for an open auction.
+        // console.log("endBlock %s currentBlock %s", _endBlock, block.number);
+        uint256 biddingPrice;
+        if (isAuctionClosed()) {
+            biddingPrice = _getCurrentPriceClosedAuction();
+        } else {
+            biddingPrice = _calcCurrentPriceOpenAuction();
+        }
+        // sanity and security check
+        if (biddingPrice == 0) {
+            revert("err:biddingPrice cant be 0");
+        } else {
+            return biddingPrice;
+        }
+    }
+
+    /// calc methods
+
+    function _getCurrentPriceClosedAuction() internal view returns (uint256 auctionPrice) {
+        return _endPrice;
+    }
+
+    function _calcCurrentPriceOpenAuction() internal view returns (uint256 auctionPrice) {
+        /// y = ax + b
+        /// => y = current auction price.
+        /// linear graph show price in Y and price is going down over time so we'll need y = -ax + b
+        /// might be easier to think about this as y = b - ax
+        /// time is the X axis here so when we start auction t=0, when we end t=delt(startBlock, endBlock)
+        /// we want to get time that is between 0 and 1 so we'll do
+        /// so this means t_current = (currentBlock - startBlock)/delta(startBlock, endBlock)
+        /// b = startPrice. b has to equal startPrice since t=0 at that point
+        /// a = (startingPrice - endPrice)
+        // currentPrice =  b - ax = startPrice - (startingPrice - endPrice) * t(0...1 only)
+        uint256 deltaInPrices = _endPrice - _startPrice;
+        console.log("deltaInPrices = %s", deltaInPrices);
+        console.log("start block %s, end block %s", _startBlock, _endBlock);
+        uint256 deltaInPriceMulCurrentTime = (deltaInPrices * (block.number - _startBlock)) / (_endBlock - _startBlock);
+        console.log("deltaInPriceMulCurrentTime = %s", deltaInPriceMulCurrentTime);
+
+        uint256 maxPriceForAuction = _startPrice;
+        console.log("maxPriceForAuction = %s", maxPriceForAuction);
+
+        uint256 currentPrice = maxPriceForAuction + deltaInPriceMulCurrentTime;
+        console.log("currentPrice = %s", currentPrice);
```

Comment:
```suggestion
        uint256 deltaInPriceMulCurrentTime = (deltaInPrices * (block.number - _startBlock)) / (_endBlock - _startBlock);
        uint256 maxPriceForAuction = _startPrice;
        uint256 currentPrice = maxPriceForAuction + deltaInPriceMulCurrentTime;
```
[Recommended to remove logging to decrease gas prices](https://ethereum.stackexchange.com/questions/110126/do-i-need-to-remove-hardhat-console-log-from-solidity-contract-before-deploying)